internal/controller/engine/engine.go

```diff
@@ -0,0 +1,411 @@
+	e.mx.RLock()
+	c, ok := e.controllers[name]
+	e.mx.RUnlock()
+
+	if !ok {
+		return errors.Errorf("controller %q is not running", name)
+	}
+
+	watchExists := make(map[schema.GroupVersionKind]bool)
+	c.mx.RLock()
+	for gvk := range c.sources {
+		watchExists[gvk] = true
+	}
+	c.mx.RUnlock()
+
+	// It's possible that we didn't explicitly stop a watch, but its backing
+	// informer was removed. This implicitly stops the watch by deleting its
+	// backing listener. If a watch exists but doesn't have an active informer,
+	// we want to restart the watch (and, implicitly, the informer).
+	a := e.infs.ActiveInformers()
+	activeInformer := make(map[schema.GroupVersionKind]bool, len(a))
+	for _, gvk := range a {
+		activeInformer[gvk] = true
+	}
+
+	// Using a map here deduplicates watches by GVK. If we're asked to start
+	// several watches for the same GVK at in the same call, we'll only start
+	// the last one.
+	start := make(map[schema.GroupVersionKind]Watch, len(ws))
+	for _, w := range ws {
+		gvk, err := apiutil.GVKForObject(w.kind, e.mgr.GetScheme())
+		if err != nil {
+			return errors.Wrapf(err, "cannot determine group, version, and kind for %T", w.kind)
+		}
+
+		// We've already created this watch and the informer backing it is still
+		// running. We don't need to create a new watch.
+		if watchExists[gvk] && activeInformer[gvk] {
+			e.log.Debug("Watch exists for GVK, not starting a new one", "controller", name, "watched-gvk", gvk)
+			continue
+		}
+
+		start[gvk] = w
+	}
+
+	// Don't take any write locks if there's no watches to start.
+	if len(start) == 0 {
+		return nil
+	}
+
+	// TODO(negz): If this blocks too much, we could alleviate it a bit by
+	// reading watches to start from a buffered channel.
+
+	// Take the write lock. This will block any other callers that want to
+	// update the watches for the same controller.
+	c.mx.Lock()
+	defer c.mx.Unlock()
+
+	// Start new sources.
+	for gvk, w := range start {
+		// The controller's Watch method just calls the StoppableSource's Start
+		// method, passing in its private work queue as an argument. This
+		// will start an informer for the watched kind if there isn't one
+		// running already.
+		//
+		// The watch will stop sending events when either the source is stopped,
+		// or its backing informer is stopped. The controller's work queue will
+		// stop processing events when the controller is stopped.
+		src := NewStoppableSource(e.infs, w.kind, w.handler, w.predicates...)
+		if err := c.ctrl.Watch(src); err != nil {
+			return errors.Wrapf(err, "cannot start watch for %q", gvk)
+		}
+
+		// Record that we're now running this source.
+		c.sources[gvk] = src
+
+		e.log.Debug("Started watching GVK", "controller", name, "watched-gvk", gvk)
+	}
+
+	return nil
+}
+
+// StopWatches stops all watches for the supplied controller, except watches for
+// the GVKs supplied by the keep argument. StopWatches blocks other operations
+// on the same controller if and when it stops a watch.
```

Comment:
StopWatches and RemoveUnwatchedInformers are used to build a higher level garbage collector that (unlike this engine) is specifically aware of XRs. See `composite/watch/watch.go`.